### DISCUSSION TITLE
Deserialize Static Data to Dynamic Data

### DISCUSSION THREAD

**Discussion by bshantam97** [eProsima/Fast-DDS#5062] [Support]
Hello All
I am trying to develop an application that will recieve a SerializedPayload created from the PubSubType of a data type defined using IDL. This is Non-Dynamic Type Data. Now the problem statement is that I want to deserialize this, on the subscription side into a dynamic Data of the same type. Can this be done ? I am working on FastDDS 2.13.2 at the moment. Please feel free to ask clarifying questions.
Thanks!

**Comment by elianalf** [comment]
Hi @bshantam97,
thanks for using Fast DDS.
You can take a look at the example of [Dynamics Types](https://github.com/eProsima/Fast-DDS/tree/2.13.x/examples/cpp/dds/DynamicHelloWorldExample). The example uses a [DynamicDataHelper](https://github.com/eProsima/Fast-DDS/blob/2.13.x/include/fastrtps/types/DynamicDataHelper.hpp) header that contains methods to print complex objects using get methods in [DynamicData.h](https://github.com/eProsima/Fast-DDS/blob/2.13.x/include/fastrtps/types/DynamicData.h). Moreover, you can take a look at [DynamicPubSubType.h](https://github.com/eProsima/Fast-DDS/blob/2.13.x/include/fastrtps/types/DynamicPubSubType.h) which contains methods to directly serialize and deserialize data.

**Reply by bshantam97** [reply]
HI @elianalf , Thank you for your response. So I have gone through the example but my requirement is that let say I am getting a serialized Non-Dynamic Type data from a DDS Framework, how would I, given the TypeInformation and the serialized packet using static data convert it to a dynamicData Representation. Please feel free to ask any clarifying questions. Thank you!

**Reply by elianalf** [reply]
If you are getting a serialized data `SerializedPayload_t`, you can use the `deserialize()` method in [DynamicPubSub.h](https://github.com/eProsima/Fast-DDS/blob/2.13.x/include/fastrtps/types/DynamicPubSubType.h#L62)

**Reply by elianalf** [reply]
So if you are looking for the specific method you can use `deserialize()`, but the example I sent to you should already work also in case the data you are sending is not a DynamicData. You should be able to retrieve a DynamicData from [on_data_available()](https://github.com/eProsima/Fast-DDS/blob/2.13.x/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldSubscriber.cpp#L113()) even if you are sending a static data.

**Reply by bshantam97** [reply]
https://github.com/eProsima/Fast-DDS/issues/2635 
@elianalf The above is the link to a similar ticket. I am getting the same backtrace. Now the issue is lets say I define a HelloWorld.idl file. This will generate HelloWorldPubSubType.cxx. Now the problem statement is that can i publish by registering HelloWorldPubSubtype, and then using DynamicPubSubtype deserialize this Static data into dynamic data representation.

**Reply by elianalf** [reply]
Yes, you can. You would need to regenerate the types with `fastddsgen` tool in the HelloWorld example with -typeobject and -replace options and use the same topic and data type in the DynamicHelloWorld example. Here the patch to modify the DynamicHelloWorld Example to be compatible with the HelloWorld Example. 

``` 
diff --git a/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldPublisher.cpp b/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldPublisher.cpp
index 567653d7a..a59be0a5c 100644
--- a/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldPublisher.cpp
+++ b/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldPublisher.cpp
@@ -94,7 +94,7 @@ bool HelloWorldPublisher::init()
         return false;
     }
 
-    topic_ = mp_participant->create_topic("DDSDynHelloWorldTopic", "HelloWorld", TOPIC_QOS_DEFAULT);
+    topic_ = mp_participant->create_topic("HelloWorldTopic", "HelloWorld", TOPIC_QOS_DEFAULT);
 
     if (topic_ == nullptr)
     {
diff --git a/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldSubscriber.cpp b/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldSubscriber.cpp
index 753a1590e..bc73ad5de 100644
--- a/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldSubscriber.cpp
+++ b/examples/cpp/dds/DynamicHelloWorldExample/HelloWorldSubscriber.cpp
@@ -166,7 +166,7 @@ void HelloWorldSubscriber::initialize_entities()
 
     //CREATE THE TOPIC
     Topic* topic = mp_participant->create_topic(
-        "DDSDynHelloWorldTopic",
+        "HelloWorldTopic",
         m_type->getName(),
         TOPIC_QOS_DEFAULT);
 
diff --git a/examples/cpp/dds/HelloWorldExample/HelloWorld.cxx b/examples/cpp/dds/HelloWorldExample/HelloWorld.cxx
index 4b415d677..ccd2ce797 100644
--- a/examples/cpp/dds/HelloWorldExample/HelloWorld.cxx
+++ b/examples/cpp/dds/HelloWorldExample/HelloWorld.cxx
@@ -28,7 +28,7 @@ char dummy;
 
 #include "HelloWorld.h"
 
-#if FASTCDR_VERSION_MAJOR > 1
+#include "HelloWorldTypeObject.h"
 
 #include <fastcdr/Cdr.h>
 
@@ -43,6 +43,8 @@ using namespace eprosima::fastcdr::exception;
 
 HelloWorld::HelloWorld()
 {
+    // Just to register all known types
+    registerHelloWorldTypes();
 }
 
 HelloWorld::~HelloWorld()
@@ -165,4 +167,3 @@ std::string& HelloWorld::message()
 // Include auxiliary functions like for serializing/deserializing.
 #include "HelloWorldCdrAux.ipp"
 
-#endif // FASTCDR_VERSION_MAJOR > 1
diff --git a/examples/cpp/dds/HelloWorldExample/HelloWorld.h b/examples/cpp/dds/HelloWorldExample/HelloWorld.h
index 5cd0ddba5..b95bcb5b1 100644
--- a/examples/cpp/dds/HelloWorldExample/HelloWorld.h
+++ b/examples/cpp/dds/HelloWorldExample/HelloWorld.h
@@ -19,11 +19,6 @@
  * This file was generated by the tool fastddsgen.
  */
 
-#include <fastcdr/config.h>
-#include "HelloWorldv1.h"
-
-#if FASTCDR_VERSION_MAJOR > 1
-
 #ifndef _FAST_DDS_GENERATED_HELLOWORLD_H_
 #define _FAST_DDS_GENERATED_HELLOWORLD_H_
 
@@ -192,4 +187,3 @@ private:
 
 
 
-#endif // FASTCDR_VERSION_MAJOR > 1
diff --git a/examples/cpp/dds/HelloWorldExample/HelloWorldPublisher.cpp b/examples/cpp/dds/HelloWorldExample/HelloWorldPublisher.cpp
index 86f3e0009..457e0bb81 100644
--- a/examples/cpp/dds/HelloWorldExample/HelloWorldPublisher.cpp
+++ b/examples/cpp/dds/HelloWorldExample/HelloWorldPublisher.cpp
@@ -61,6 +61,9 @@ bool HelloWorldPublisher::init(
         return false;
     }
 
+    type_->auto_fill_type_information(false);
+    type_->auto_fill_type_object(true);
+
     //REGISTER THE TYPE
     type_.register_type(participant_);
 ``

**Reply by elianalf** [reply]
Additionally, we are working on the new major release of Fast DDS, v3.0.0, which is now Master. In the new version we have refactored the dynamic types with improvements and fixes, so I would suggest you switch to Fast DDS Master. It is currently unstable but will be released soon. You should change the name of some APIs but it will probably fix the segmentation fault.

**Reply by bshantam97** [reply]
@elianalf THank you for your response. So this is something I have done. I constructu my dynamic Type the following way
registerXXXTYpe()
auto type_id = GetXXXTypeId(true) // Complete Identifier
auto type_obj = GetXXXTypeObject(true)
ypeObjectFactory::get_instance()->build_dynamic_type(topic_name, type_id, type_obj);
Now I have another DDS Framework that just sends me static data a SerializedPayload_t to be specific. I am able to process that. Can I do something like the following 
DynamicPubSubType->deserialize(SerializedPayload_t(For Static Data), dynamicData.get()) and then automatically populate it ?
From what I have tried till now, this fails. 
Thank you. Please let me know if I can clarify this any further. At the moment we have a requirement to stay on the FastDDS 2.13.2.
In addition I am alread using the -typeobject and -replace with fastddsgen, to generate my types. I think the core question is can this static data recieved be automatically mapped to its associated Dynamic Data Representation in C++

**Reply by elianalf** [reply]
It should work, so I think you should provide a reproducer with all the files we need to reproduce it.
